third_party/pcsc-lite/naclport/js_client/src/reader-tracker-through-pcsc-api.js

```diff
@@ -0,0 +1,435 @@
+ * race-free).
+ * @param {!API} api
+ * @param {!API.SCARDCONTEXT} sCardContext
+ * @param {!Array.<!API.SCARD_READERSTATE_OUT>} previousReaderStatesOut
+ * @return {!goog.Promise}
+ * @private
+ */
+ReaderTrackerThroughPcscApi.prototype.makeReaderStatesChangePromise_ = function(
+    api, sCardContext, previousReaderStatesOut) {
+  /** @type {!goog.promise.Resolver} */
+  const promiseResolver = goog.Promise.withResolver();
+
+  const readerStatesIn = goog.array.map(
+      previousReaderStatesOut,
+      function(readerStateOut) {
+        // Clear the upper 16 bits that correspond to the number of events that
+        // happened during previous call (see the SCardGetStatusChange docs for
+        // the details).
+        const currentState = readerStateOut['event_state'] & 0xFFFF;
+        return new API.SCARD_READERSTATE_IN(
+            readerStateOut['reader_name'], currentState);
+      });
+  // Add a magic entry that tells PC/SC to watch for the reader adding/removing
+  // events too (see the SCardGetStatusChange docs for the details).
+  readerStatesIn.push(new API.SCARD_READERSTATE_IN(
+      '\\\\?PnP?\\Notification', API.SCARD_STATE_UNAWARE));
+
+  this.logger_.fine(
+      'Waiting for the reader statuses change from PC/SC with the following ' +
+      'data: ' + GSC.DebugDump.dump(readerStatesIn) + '...');
+
+  api.SCardGetStatusChange(
+      sCardContext,
+      READER_STATUS_QUERY_TIMEOUT_MILLISECONDS,
+      readerStatesIn).then(function(result) {
+    result.get(function(readerStatesOut) {
+      this.logger_.fine('Received a reader statuses change event from PC/SC: ' +
+                        GSC.DebugDump.dump(readerStatesOut));
+      promiseResolver.resolve();
+    }, function(errorCode) {
+      if (errorCode == API.SCARD_E_TIMEOUT) {
+        this.logger_.fine('No reader statuses changes were reported by PC/SC ' +
+                          'within the timeout');
+        promiseResolver.resolve();
+      } else if (errorCode == API.SCARD_E_UNKNOWN_READER) {
+        this.logger_.warning(
+            'Waiting for the reader statuses changes from PC/SC finished ' +
+            'unsuccessfully due to removal of the tracked reader');
+        promiseResolver.resolve();
+      } else {
+        promiseResolver.reject(new Error(
+            'Failed to wait for the reader statuses change from PC/SC with ' +
+            'error code ' + GSC.DebugDump.dump(errorCode)));
+      }
+    }, this);
+  }, function(error) {
+    promiseResolver.reject(new Error(
+        'Failed to get the reader statuses from PC/SC: ' + error));
+  }, this);
+
+  return promiseResolver.promise;
+};
+
+/**
+ * Updates the result reported by this instance, given the reader states
+ * obtained from a call to the SCardGetStatusChange function.
+ * @param {!Array.<!API.SCARD_READERSTATE_OUT>} readerStates
+ * @private
+ */
+ReaderTrackerThroughPcscApi.prototype.updateResultFromReaderStates_ = function(
```

Comment:
nit: I would put this function right before makeReaderStatesChangePromise_ since it is called right before that one.

Reply:
Done